src/controllers/scripting/controllerscriptenginebase.h

```diff
@@ -49,21 +63,67 @@ class ControllerScriptEngineBase : public QObject {
     void scriptErrorDialog(const QString& detailedError, const QString& key, bool bFatal = false);
     void logOrThrowError(const QString& errorMessage);
 
+#ifdef MIXXX_USE_QML
+    inline void setQMLMode(bool qmlFlag) {
+        m_bQmlMode = qmlFlag;
+    }
+    inline void setErrorsAreFatal(bool errorsAreFatal) {
+        m_bErrorsAreFatal = errorsAreFatal;
+    }
+#endif
+
     bool m_bDisplayingExceptionDialog;
+#ifdef MIXXX_USE_QML
+    bool m_bErrorsAreFatal;
+#endif
     std::shared_ptr<QJSEngine> m_pJSEngine;
 
     Controller* m_pController;
     const RuntimeLoggingCategory m_logger;
 
     bool m_bAbortOnWarning;
 
+#ifdef MIXXX_USE_QML
+    bool m_bQmlMode;
+#endif
     bool m_bTesting;
 
+#ifdef MIXXX_USE_QML
+  private:
+    static inline std::shared_ptr<TrackCollectionManager> s_pTrackCollectionManager;
+
+    /// Pause the GUI main thread. Pause is required by rendering
+    /// thread (https://doc.qt.io/qt-6/qquickrendercontrol.html#sync). This
+    /// offscreen render thread to pause the main "GUI thread" for onboard
+    /// screens
+    /// The documentation isn't completely clear about this, but after
+    /// testing, it appears that the "GUI main thread" is the thread where the QML
+    /// engine leaves in (also the main thread if we were using a
+    /// QMLApplication, which isn't the case here)
+    QWaitCondition m_isPausedCondition;
+    QMutex m_pauseMutex;
+    bool m_isPaused{false};
+    bool m_canPause{false};
+
+  public slots:
+    bool pause();
+    void resume();
+    void setCanPause(bool canPause);
```

Comment:
I don't really get how this is supposed to work. From what I understood, this was supposed to pause the QML GUI thread, but it pauses the controller engine?

Reply:
Yes, the QML engine is running as the controller engine . The doc is unclear and at first, I thought the "GUI thread" was the main GUI thread, turns out it means the QML thread. This is required because the rendering is done in a different thread (main or controller rendering depending the version)

Reply:
So do I understand correctly that we now have a QJSEngine and a QQMLEngine running in the same thread?

Reply:
No, we have either a `QJSEngine` or a `QQMLEngine` (which inherit `QJSEngine`)

Reply:
Okay, so the JS code of the mapping and the QtQuick code is running in the same engine? That would imply that controller IO handling is also blocked while rendering the frame. That would likely cause problems with expensive UIs / slow devices. Does that make sense?

Reply:
>  That would imply that controller IO handling is also blocked while rendering the frame.

No, this is why the rendering runs in its thread. The only time the controller thread is blocked is during the scene graph sync (the "pause" feature) which is about ~1us (can be tested with ScopedTimer now), actual rendering and data communication has no impact on controller thread and IO.